### ISSUE TITLE
book page 68, some_data_prepared dimension not right(with my own solution, need verified)

### ISSUE THREAD

**Issue by MLShooter** [ageron/handson-ml#207]
i found the code some_data_prepared = full_pipeline.transform(some_data) somehow will not generate a length 16 vector to feed the predictor, it is due to 5 samples can not cover all the categories, so the sparse matrix output by the cat_pipeline will be shorten automatically. the following is my solution.

step 1: create a dictionary of the cat_attribs
```
#handling text and categorical attributes
from sklearn.preprocessing import LabelEncoder
encoder = LabelEncoder()
housing_cat = housing["ocean_proximity"]
housing_cat_encoded = encoder.fit_transform(housing_cat)
hot_dic = {}
cat_idx = 0
for cat in encoder.classes_:
    tmp_ar = [0] * len(encoder.classes_)
    tmp_ar[cat_idx] = 1
    hot_dic[cat] = tmp_ar
```

step2: use a customized LabelBinarizer
```
class CustomLabelBinarizer(BaseEstimator, TransformerMixin):
    def __init__(self, hot_dic):
        self.hot_dic = hot_dic
    def fit(self, X, y=None):
        return self
    def transform(self, X, y=None):
        tmp_l = []
        for loc in range(len(X)):
            tmp_l.append(self.hot_dic[X[loc][0]])
        return np.array(tmp_l)

```
step3: give hot_dic to the cat_pipeline
```
cat_pipeline = Pipeline([
    ('selector', DataFrameSelector(cat_attribs)),
    ('label_binarizer', CustomLabelBinarizer(hot_dic)),
])

```
that's all i need to modify to get all that working, would you please check whether there'are some better 
methods ?

**Comment by ageron** [comment]
Hi @MLShooter ,

I'm glad you managed to work around the problem.  There is still no great way to perform one-hot encoding on the input feature in Scikit-Learn.  The `LabelEncoder` class is only meant to encode labels, not input features: it only handles 1D arrays, and if you use it for input features (one feature at a time), it won't work in Pipelines.  Your solution is one way to work around the limitation.
In Scikit-Learn 0.20 (next version), they initially planned to have a `CategoricalEncoder` class for this purpose, but it seems that there will be two classes instead: `OrdinalEncoder` and `OneHotEncoder`(see https://github.com/scikit-learn/scikit-learn/issues/10521).

Hope this helps,
Aurélien

**Comment by MLShooter** [comment]
@ageron Thank you very much~
